Remove form entirely #129 (#142)
enonic/app-oidc-idprovider: docs/index.adoc

==== Virtual Host Mapping configuration

Edit the configuration file `com.enonic.xp.web.vhost.cfg`, and set the new user store to your virtual host. (See https://developer.enonic.com/docs/xp/stable/deployment/vhosts[Virtual Host Configuration] for more information).

[source,properties]
----
mapping.example.host = example.com
mapping.example.source = /
mapping.example.target = /portal/master/mysite
mapping.example.idProvider.myoidcprovider = default
----

Here as well, the ID provider name must match: see the last line.

{zwsp} +

[[idp_config]]
=== Configure the ID Provider

As of v2.0.0, the config form in the users app (`idprovider.xml`) has been removed. The settings to configure the id provider must instead be entered in a .CFG file: com.enonic.app.oidcidprovider.cfg.

> In this .cfg file, all ID providers that use this app are configured. Each ID provider gets a idprovider.<idprovidername> namespace, which prefixes the config fields. Add keys to that on the format `idprovider.[idprovidername].[configkey(s)] = [value]`. The ID provider name is set in the Users app in XP. For example, setting the `forceEmailVerification` rule config field with a value of `true` for an ID provider named myoidcprovider, will look like this: `idprovider.myoidcprovider.rules.forceEmailVerification=true`

Reviewer: ```suggestion
This config file should contain config fields for all OIDC ID providers in the format `idprovider.[idprovidername].[configkey] = [value]`, where `idprovidername` should exactly match ID provider name from the Users app. For example, setting the `forceEmailVerification` rule config field with a value of `true` for an ID provider named `myoidcprovider`, will look like this: `idprovider.myoidcprovider.rules.forceEmailVerification=true`
```

idprovider.<idprovidername>.additionalEndpoints.0.url=  (string, required)

# Optional 'endSession' namespace
idprovider.<idprovidername>.endSession.url=  (string, required)
idprovider.<idprovidername>.endSession.idTokenHintKey=  (string, optional)
idprovider.<idprovidername>.endSession.postLogoutRedirectUriKey=  (string, optional)

# Optional 'additionalParameters' namespace
# Is an array, that uses index (integer, starts from 0) to set items with 'key' and 'value' fields
idprovider.<idprovidername>.endSession.additionalParameters.0.key=  (string, required)
idprovider.<idprovidername>.endSession.additionalParameters.0.value=  (string, required)

idprovider.<idprovidername>.clientId=  (string, required)
idprovider.<idprovidername>.clientSecret=  (string, required)

idprovider.<idprovidername>.mappings.displayName=  (string, required)
idprovider.<idprovidername>.mappings.email=  (string, required)
idprovider.<idprovidername>.mappings.defaultGroups=  (space separated group keys, optional)

idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, required)
----

{zwsp} +

[[fields_in_config]]
==== Fields in the config

> * `autoinit`: Automatic initialization. If `com.enonic.app.oidcidprovider.cfg` contains `autoinit=true`, then during startup this app will look all idprovider names declared in the file and create them (if the app names don't already exist) with those settings. For example, `idprovider.myfirstidp.someKey=someValue` and `idprovider.anotheridp.anotherKey=anotherValue` will declare two idproviders named `myfirstidp` and `anotheridp`.

Reviewer: ```suggestion
* `autoinit`: Automatic initialization. If the config file contains `autoinit=true`, then during startup this app will automatically create ID providers for all settings declared in the file, if they don't already exist. For example, `idprovider.myfirstidp.someKey=someValue` and `idprovider.anotheridp.anotherKey=anotherValue` will declare two idproviders named `myfirstidp` and `anotheridp`.
```

  ** `url`: End session URL. Value of `end_session_endpoint` in your OpenID Provider <<#information_required_for_config, configuration>>.
  ** `idTokenHintKey`: ID Token Hint parameter name. Value of `id_token_hint` in your OpenID Provider <<#information_required_for_config, configuration>>.
  ** `postLogoutRedirectUriKey`: Post logout redirect URI parameter name. Value of `post_logout_redirect_uri` in your OpenID Provider <<#information_required_for_config, configuration>>.
  ** `additionalParameters...` (`idprovider.<idprovidername>.endSession.additionalParameters...` namespace): Additional OIDC Front-Channel Logout. This `additionalParameters` namespace _requires an array_ immediately below it, with two required fields below each item in the array. Array starts counting on `0`, so for example, `idprovider.<idprovidername>.endSession.additionalParameters.0.key` sets the `key` for the first item, while `idprovider.<idprovidername>.endSession.additionalParameters.1.value` sets `value` for the second one, etc.
    *** `key`
    *** `value`

**Client**

> * `clientId`: Client identifier, received ot generated during the <<#authorization_server, creation of your Client/Application>> (this application uses the authentication method "client_secret_post" for the token retrieval).

Reviewer: ```suggestion
* `clientId`: Client identifier, received or generated during <<#authorization_server, creation of your Client/Application>> (this application uses the authentication method "client_secret_post" for token retrieval).
```

Example: **Azure AD**

* End Session URL: https://login.microsoftonline.com/{TenantID}/oauth2/logout
* Post Logout Redirect URI parameter name: `post_logout_redirect_uri`

{zwsp} +

Example: **Difi/ID-porten/VER1**

* End Session URL: https://oidc-ver1.difi.no/idporten-oidc-provider/endsession
* ID Token Hint parameter name: `id_token_hint`
* Post Logout Redirect URI parameter name: `post_logout_redirect_uri`

{zwsp} +

== Upgrading from version 1.x
Version `2.x` of this app introduces a breaking change: all configuration now happens in `com.enonic.app.oidcidprovider.cfg` instead of editing a form in the Users app. So no configuration will be stored in / read from the data layer anymore.

> The following upgrade descriptions use information you can find by editing your `1.x` ID provider app in XP:

Reviewer: ```suggestion
The following upgrade description assumes you currently have an earlier (`1.x`) version of OIDC ID provider app installed:
```

image:idprovider-migration.png[title="(1) - ID provider name, (2) - pencil button to open form for edit", width=500px]

When editing an ID provider in the Users app, the ID provider name (1) is found (or edited) directly below the displayName field, and in previous versions of this app the configuration form was opened by clicking the pencil (2) icon on the ID provider application.

=== Migrate the configuration

> Create `com.enonic.app.oidcidprovider.cfg` in your `/config` folder. Then in XP, enter the Users app and edit the ID provider. Note the ID provider name (without the leading slash), and click the edit-form icon to view the configuration entered in the old app. Transfer the values into the `.cfg` file, the way it's specified above (the order may vary).

Reviewer: ```suggestion
Create `com.enonic.app.oidcidprovider.cfg` in your XP instance's `config` folder. Then in the XP backoffice, open the Users app and edit the ID provider. Note the ID provider name (without the leading slash), and click the edit-form icon to view the configuration entered in the old app. Transfer the values into the `.cfg` file, the way it's specified above (the order may vary).
```
